src/utils_ovs.c

```diff
@@ -0,0 +1,1287 @@
+  ovs_json_reader_free(jreader);
+  pthread_exit((void *)0);
+  return ((void *)0);
+}
+
+/* EVENT worker thread.
+ * Perform task based on incoming events. This
+ * task can be done asynchronously which allows to
+ * handle OVD DB callback like 'init_cb'.
+ */
+static void *ovs_event_worker(void *arg) {
+  int ret = 0;
+  ovs_db_t *pdb = (ovs_db_t *)arg;
+  struct timespec ts;
+
+  while (pdb->event_thread.value != OVS_DB_EVENT_TERMINATE) {
+    /* wait for an event */
+    clock_gettime(CLOCK_REALTIME, &ts);
+    ts.tv_sec += (OVS_DB_EVENT_TIMEOUT);
+    ret = pthread_cond_timedwait(&pdb->event_thread.cond,
+                                 &pdb->event_thread.mutex, &ts);
+    if (!ret) {
+      /* handle the event */
+      OVS_DEBUG("handle event %d", pdb->event_thread.value);
+      switch (pdb->event_thread.value) {
+      case OVS_DB_EVENT_CONN_ESTABLISHED:
+        if (pdb->cb.post_conn_init)
+          pdb->cb.post_conn_init(pdb);
+        break;
+      case OVS_DB_EVENT_CONN_TERMINATED:
+        if (pdb->cb.post_conn_terminate)
+          pdb->cb.post_conn_terminate();
+        break;
+      default:
+        OVS_DEBUG("unknown event received");
+        break;
+      }
+    } else if (ret == ETIMEDOUT) {
+      /* wait timeout */
+      OVS_DEBUG("no event received (timeout)");
+      continue;
+    } else {
+      /* unexpected error */
+      OVS_ERROR("pthread_cond_timedwait() failed");
+      break;
+    }
+  }
+
+thread_exit:
+  OVS_DEBUG("event thread has been completed");
+  pthread_exit((void *)0);
+  return ((void *)0);
+}
+
+/* Stop EVENT thread */
+static int ovs_db_event_thread_stop(ovs_db_t *pdb) {
+  ovs_db_event_post(pdb, OVS_DB_EVENT_TERMINATE);
+  if (pthread_join(pdb->event_thread.tid, NULL) != 0)
+    return (-1);
+  pthread_mutex_unlock(&pdb->event_thread.mutex);
+  pthread_mutex_destroy(&pdb->event_thread.mutex);
+  return (0);
+}
+
+/* Stop POLL thread */
+static int ovs_db_poll_thread_stop(ovs_db_t *pdb) {
+  ovs_db_poll_terminate(pdb);
+  if (pthread_join(pdb->poll_thread.tid, NULL) != 0)
+    return (-1);
+  pthread_mutex_destroy(&pdb->poll_thread.mutex);
+  return (0);
+}
+
+/*
+ * Public OVS DB API implementation
+ */
+
+ovs_db_t *ovs_db_init(const char *node, const char *service,
+                      ovs_db_callback_t *cb) {
+  pthread_mutexattr_t mutex_attr;
+  ovs_db_t *pdb = NULL;
+
+  /* allocate db data & fill it */
+  if ((pdb = calloc(1, sizeof(*pdb))) == NULL)
+    return (NULL);
+
+  /* node cannot be unset */
+  if (node == NULL || strlen(node) == 0)
+    return (NULL);
+
+  /* store the OVS DB address */
+  sstrncpy(pdb->node, node, sizeof(pdb->node));
+  if (service != NULL)
+    sstrncpy(pdb->service, service, sizeof(pdb->service));
+
+  /* setup OVS DB callbacks */
+  if (cb)
+    pdb->cb = *cb;
+
+  /* prepare event thread */
+  pthread_cond_init(&pdb->event_thread.cond, NULL);
+  pthread_mutex_init(&pdb->event_thread.mutex, NULL);
+  pthread_mutex_lock(&pdb->event_thread.mutex);
```

Comment:
Is this mutex held on purpose when returning from this function? If so, please add this to this function's documentation.